src/mlpack/core/optimizers/pso/pso_impl.hpp

```diff
@@ -0,0 +1,106 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_PSO_PSO_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_PSO_PSO_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "pso.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template<typename VelocityVectorType>
+PSOType<VelocityVectorType>::PSOType(const size_t lambda,
+                                  const size_t dimension,
+                                  const double interiaWeight,
+                                  const double cognitiveAcceleration,
+                                  const double socialAcceleration,
+                                  const size_t maxIterations,
+                                  const double tolerance,
+                                  const VelocityVectorType& velocityType) :
+    lambda(lambda),
+    dimension(dimension),
+    interiaWeight(interiaWeight),
+    cognitiveAcceleration(cognitiveAcceleration),
+    socialAcceleration(socialAcceleration),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    velocityType(velocityType)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename VelocityVectorType>
+template<typename DecomposableFunctionType>
+double PSOType<VelocityVectorType>::Optimize(
+    DecomposableFunctionType& function, arma::mat& iterate)
+{
+  // Randomly initialize the particle position and velocity.
+  // Following a heuristic, the swarm size is set to 2 * dimension.
+  lambda = 2 * dimension;
+  particlePosition.randu(dimension);
+  particleVelocity.randu(dimension);
+  bestParticlePosition = particlePosition;
+  bestSwarmPosition = particlePosition;
+
+  // Convenient variables to check if there's an improvement.
+  double currentObjective;
+  double lastObjectiveIndividual = DBL_MAX;
+  double lastObjectiveGlobal = DBL_MAX;
+
+  // Start iterating.
+  for (size_t i = 0; i < maxIterations; ++i)
+  {
+    for (size_t k = 0; k < dimension; ++k)
+    {
+      // Calculate the objective function.
+      currentObjective = function.Evaluate(particlePosition);
+
+      // Check if the current position is an individual best.
+      if (currentObjective < lastObjectiveIndividual)
+      {
+        bestParticlePosition = particlePosition;
+        lastObjectiveIndividual = currentObjective;
+      }
+    }
+
+    // Check if the current position is a global best.
+    if (lastObjectiveIndividual < lastObjectiveGlobal)
+    {
+      bestSwarmPosition = particlePosition;
+      lastObjectiveGlobal = lastObjectiveIndividual;
+    }
+
+    // Update velocity for each particle.
+    velocityType.Update(particlePosition,
+      particleVelocity, bestParticlePosition,
+      bestSwarmPosition, interiaWeight,
+      cognitiveAcceleration, socialAcceleration, dimension);
+
+    // Update position for each particle.
+    particlePosition = particleVelocity;
```

Comment:
This is incorrect. In the paper the position is updated like this:
```
particlePosition = particlePosition + particleVelocity;
```
However, when I tried this here `particleVelocity` gets extremely small (around `1e-200`) in a few iterations. So, adding it to `particlePosition` does not have any effect. This leads the objective function to return the same value in each iteration, hence it does not get minimized.

I have spent some time trying to fix this, but haven't been able to do so yet. However, I suspect this could be because of the following two reasons:

- In the other implementations I found online ([swarmops - Python](https://github.com/Hvass-Labs/swarmops/blob/master/swarmops/PSO.py#L346), [C](https://github.com/kkentzo/pso/blob/master/pso.c#L380)), there is a boundary enforcement after each update. I haven't done this as this would be a feature for constrained optimization (?).
- Currently, the particle position and velocity is being initialized with uniform random values. This is also how its done in the other implementations I could find. But, this may be causing the particle velocity to 'die out'. There is a [question regarding this on ResearchGate](https://www.researchgate.net/post/How_to_initialize_the_position_and_velocity_of_the_particle_swarm_optimization_technique_for_PV_model) which links to some papers, but I haven't gone through any of those.

I would really appreciate if someone could help me figure out how to resolve this issue.

Reply:
I don't think we could get around using a bound here, even if we find a good initialization. However, I like the idea to test different initalization methods, one idea we could test is, to use CNE (https://github.com/mlpack/mlpack/tree/master/src/mlpack/core/optimizers/cne) or CMAES (https://github.com/mlpack/mlpack/tree/master/src/mlpack/core/optimizers/cmaes) to initialize the velocities. Let me know what you think.